pkg/repository/prisma/workflow_run.go

```diff
@@ -838,6 +839,221 @@ func (w *workflowRunEngineRepository) GetScheduledChildWorkflowRun(ctx context.C
 
 	return w.queries.GetScheduledChildWorkflowRun(ctx, w.pool, childParams)
 }
+
+func (w *workflowRunEngineRepository) PopWorkflowRunsCancelInProgress(ctx context.Context, tenantId, workflowId string, maxRuns int) ([]*dbsqlc.WorkflowRun, []*dbsqlc.WorkflowRun, error) {
+	ctx, span := telemetry.NewSpan(ctx, "queue-by-cancel-in-progress")
+	defer span.End()
+
+	tx, commit, rollback, err := sqlchelpers.PrepareTx(ctx, w.pool, w.l, 15000)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	defer rollback()
+
+	// place a FOR UPDATE lock on queued and running workflow runs to prevent concurrent updates
+	allToProcess, err := w.queries.LockWorkflowRunsForQueueing(ctx, tx, dbsqlc.LockWorkflowRunsForQueueingParams{
+		Tenantid:   sqlchelpers.UUIDFromStr(tenantId),
+		Workflowid: sqlchelpers.UUIDFromStr(workflowId),
+	})
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not lock workflow runs for queueing: %w", err)
+	}
+
+	// group each workflow run by concurrency key
+	keyToWorkflowRuns := make(map[string][]*dbsqlc.WorkflowRun)
+
+	for _, row := range allToProcess {
+		key := row.ConcurrencyGroupId.String
+
+		if _, ok := keyToWorkflowRuns[key]; !ok {
+			keyToWorkflowRuns[key] = make([]*dbsqlc.WorkflowRun, 0)
+		}
+
+		keyToWorkflowRuns[key] = append(keyToWorkflowRuns[key], row)
+	}
+
+	allToCancel := make([]*dbsqlc.WorkflowRun, 0)
+	allToStart := make([]*dbsqlc.WorkflowRun, 0)
+	cancelIds := make([]pgtype.UUID, 0)
+
+	for _, toProcess := range keyToWorkflowRuns {
+		runningWorkflowRuns := make([]*dbsqlc.WorkflowRun, 0, len(toProcess))
+		queuedWorkflowRuns := make([]*dbsqlc.WorkflowRun, 0, len(toProcess))
+
+		for _, row := range toProcess {
+			if row.Status == dbsqlc.WorkflowRunStatusRUNNING {
+				runningWorkflowRuns = append(runningWorkflowRuns, row)
+			} else if row.Status == dbsqlc.WorkflowRunStatusQUEUED {
+				queuedWorkflowRuns = append(queuedWorkflowRuns, row)
+			}
+		}
+
+		// iterate over the running workflow runs and cancel them
+		toCancel := max(0, len(runningWorkflowRuns)+len(queuedWorkflowRuns)-maxRuns)
```

Comment:
do we need to consider cancelling?

Reply:
I don't believe so, as those no longer contribute to the maxRuns count.